Upgrade tgrs

psmithshine opened the pull request:
Upgrade `client`, `server`, `shared`, and `integration` packages to support and run Node v24. Local development and build is faster 🚀 across each package (upgrade some builders/bundlers) and the overall app's runtime (Node) is no longer deprecated!! 🔒

We also upgraded Typescript to latest version (v5.9.3) across the board, so we're feeling **very** modern 👔

## Changelog

### Server - Node v24 🔥
Upgrade Apollo + GraphQL packages to their latest versions to be compatible with Node v24. Other packages such as `concurrently` and `tsx` are there to support running the server locally in Node v24 context. Significant re-writes for the server were necessary as the API for Apollo has changed over the major version bumps and changes. The Datasources was re-written to better complement the setup for the Server. The `tsconfig` updates allows for the Server to read the new module syntax which `shared` package compiles, and brings it into the more modern Typescript era. 

### Client - Lets get [Vite](https://www.collinsdictionary.com/dictionary/french-english/vite)  ⚡️
Now using [ViteJS](https://vitejs.dev/)  build tool as a replacement of react-scripts for compiling and running the app locally, as well as building it for production. The react team doesn't even recommend using their own react-scripts tool anymore, and the community much prefer Vite over react-scripts - so it was time! You will find that Vite is significantly quicker to run localhost and to build the app.

index.html uses the new Vite-friendly syntax, like ridding the %PUBLIC_URL% stuff for project-relative paths
adding a vite.config.ts, for a basic Vite build configuration to work with our React app
package.json had the following updates:
- `type: module` for Vite to function
- tsconfig.json had some of the most important changes, to deal with Node v24 updates + Vite.
- a big part of these changes were supporting Vite's HMR (Hot Module Reloading) features, which require "ESNext" Module code (some latest ECMAScript standard, but not pegged to a specific version like ES2019, ES2020)
- we also employ the same codegen CLI that the server uses, for consistency
- another part of these changes + effort had to do with being able to deal with the Shared/ folder's Typescript compiled code -> requiring shared's `tsconfig` to also have updated compilerOptions -> specifically having module and target be ES6.
- We have also moved to vitest, a testing framework with almost 100% parity to Jest, but using the Vite engine ⏩ 

### Integration - smarter port resolution, feat Express instead of Stubby 🚂 
- Lightweight & maintainable: Switched from Stubby (deprecated/unmaintained) to Express for the mock REST server
- Declarative route configuration: New mockData.ts with type-safe route definitions (MockRoute interface)
- Smarter port resolution: Port configuration from env file: Integration environment now reads messageServerUrl from packages/server/env.integration.json to determine which port to use for the mock server

This PR uses Zod to validate env variables, Fixes #15

psmithshine committed:
make current TGRS work
- update tsconfig target and packages in shared
- update packages in server

psmithshine committed:
- make mesage be a simple "Hello", instead of using "get"

psmithshine committed:
🛰️ 🛰️ 🛰️ upgrade to Node v20 in server 🛰️ 🛰️ 🛰️
- upgrade server packages necessary to support Node v20, and remove some deprecated packages
- adopt new Apollo Server API from Apollo v4.9 (jumped two major versions), for much of the setup of the local standalone server
- use latest tsconfig for Node v20
- use more sensical Datasources sertup (to also complement newer Apollo API)

psmithshine committed:
- add @parcel/watcher package to support dep of latest GraphQL codegen cli package version

psmithshine committed:
- update lambda code to new api

psmithshine committed:
- prettier format with @ts-ignore, for correct request context parsing

psmithshine committed:
- update shared deps and tsconfig to prep for client's Vite overhaul

psmithshine committed:
⚛ ⚛ ⚛ upgrade to Node v20 and setup Vite in client ⚛ ⚛ ⚛
👍 Node v20 Upgrades 👍
- update package versions to support latest node
- use same GraphQL codegen tooling as Server for consistent code generation across BE/FE, + with codegen.yaml file
⚡️ Vite Upgrades ⚡️
- new Vite build + dev server tool - the speed!! 🚀
- supporting Vite packages, vite config, index.html setup, index.html placement, tsconfig setup, start and build scripts updates
- remove react-scripts and any react scripts setup 🤮

psmithshine committed:
- package.json giving the Node version, and preventing tools other than yarn
- update node for nodenv tool

psmithshine committed:
made tests work as-is, because it seems the webserver isnt contacting the correct server EP, so we're getting error on requesting from server

bent commented on packages/client/env/localhost-4000.json:
If the port number of the server has changed, perhaps we should change the name of this file to reflect that?

psmithshine replied:
yeah for sure. I'm happy to keep it 4000, i actually just changed it temporarily so as to not interfere with the other project's GraphQL playground queries which were saved (browser saved queries according to that localhost:4000).

bent replied:
OK, I'm still seeing the server URI as being port 2000, what's the plan here?

psmithshine replied:
done, changed back to 4000!

bent commented on packages/client/package.json:
```diff
@@ -2,33 +2,37 @@
     "start": "./scripts/start.sh",
     "build": "./scripts/build.sh",
     "test": "./scripts/test.sh",
-    "codegen": "./scripts/codegen.sh"
```
Does this mean we can get rid of `./scripts/codegen.sh` now?

psmithshine replied:
removed

bent commented on packages/client/codegen.yml:
```diff
@@ -0,0 +1,9 @@
+schema: ../server/src/schema.graphql
+generates:
+  ./src/shared/__generated__/:
```
Can we switch this to `./src/__generated__`?

psmithshine replied:
for sure - just wondering, is this because its best/customary to have further code written deeper in src to be accessing stuff from parent, rather than a reach-n-grab-to-here-from-there free-for-all?

bent replied:
This is generally the approach I've used on other projects. There's a case to be made for moving it under `shared`, but on non-trivial projects I usually also have an eslint rule to [not import from inside modules](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-internal-modules.md). This means `shared` needs to re-export everything under `shared/__generated__`, which can create namespace clashes and makes it harder for people to know whether they're dealing with generated code or not.

bent commented on packages/client/src/App.tsx:
```diff
@@ -1,18 +1,17 @@
 function App() {
-  const { data, loading, error } = useQuery<AppQuery, AppQueryVariables>(
-    gql`
+  const { data, loading, error } = useQuery(
```
I'm reluctant to remove type parameters because basically the types become `any` if we do. The types still exist in `./shared/__generated__/graphql`, it's just now they're called `AppQueryQuery` and `AppQueryQueryVariables`, as GraphQL Codegen uses a different naming convention by default. As these names are a bit clunky, I'm fine for you to rename `AppQuery` to just `App` so that the generated names are the same as what they were previously.

psmithshine replied:
Ah you're right - I didn't actually fully finish updating this part properly.
The type params actually don't need to be in there anymore when using the codegend `graphql` fn (as opposed to `gql` fn that i still have in there, my bad!). This `graphql` fn matches to the passed-in query string arg, so it is able to give you input and return types automagically. I will include this in my updates, but you can see it'll look like this:<img width="918" alt="Screenshot 2024-07-09 at 12 23 16 pm" src="https://github.com/shinesolutions/tgrs/assets/124022076/f5999904-e691-4bee-b550-3bb5a2c56230">

bent commented on packages/integration/package.json:
```diff
@@ -17,11 +17,14 @@
     "startStubby": "stubby --data ./src/stubbyData.yml"
+  },
+  "devDependencies": {
+    "ts-node": "^10.9.2"
```
Can we just remove this now that we're using `tsx` instead?

bent committed:
Got integration tests partially going again

bent committed:
Remove server tests

bent committed:
Remove client tests

bent committed:
Got full integration test going again

bent committed:
Re-added userRequired directive to schema

bent commented:
I have pushed some commits to this branch to get the integration environment and tests running properly again, and get the auth directive in-place again. Note that I have removed the client and server tests because they weren't passing with your changes, and they're not really something I use on client sites anyway (I mostly rely on integration tests).

psmithshine committed:
- remove ts-node (using tsx instead)
- use the generated `graphql` function, rather than `gql`, so we get free typings from codegend code without needing to explicitly specify them

psmithshine committed:
- update generates config path to just be into src/ and update codegen import paths accordingly
- remove codegen script

psmithshine committed:
- remove express.ts module as we now have more modern `@as-integrations/aws-lambda package`
- also remove associated comment from index.ts module

bent commented on packages/integration/package.json:
```diff
@@ -11,17 +11,18 @@
     "server": "*",
     "stubby": "^5.0.0",
-    "ts-node": "^9.0.0",
-    "typescript": "^4.0.3"
```
Any particular reason we need to use a fixed version of typescript?

psmithshine replied:
you're right, have updated to allow variable version

psmithshine committed:
- keep port as 4000

psmithshine committed:
- dont fix Typescript version

bent committed:
Upgrade Ubuntu version

bent committed:
Merge branch 'upgrade-tgrs' of https://github.com/shinesolutions/tgrs into upgrade-tgrs

bent committed:
Upgrade Cypress Github Action

psmithshine committed:
- set not-exact version

psmithshine committed:
- add cypress as devDep

psmithshine committed:
remove cypress tests

psmithshine committed:
Update build.yml

psmithshine committed:
Update build.yml

psmithshine committed:
Update build.yml

psmithshine committed:
Update build.yml

psmithshine committed:
- Introduced a new mock server using Express that allows for easy configuration of routes, like Stubby did previously
- swapped mockData for mockData.ts to define mock routes with request and response specifications.
- Implemented `startMockServer.ts` to initialize the server and handle incoming requests based on the defined routes.
- removed old Stubby stuff, in favour of Express

psmithshine committed:
use startStubby instead

psmithshine committed:
add back in yarn installs

psmithshine committed:
try previous cypress version

psmithshine committed:
upgrade cyress again

psmithshine committed:
try to run with install: false

psmithshine committed:
upgrade jsonwebtoken package

psmithshine committed:
yarn install and update node version

psmithshine committed:
- use latest TS version across repo
- use modern jose instead of jsonwebtoken for jwt stuff across repo
- use more modern tsup to build server, instead of tsc
- upgrade node/types across repo too

psmithshine committed:
upgrade typescript to version 5.9.3 across all packages

bent commented on packages/integration/package.json:
```diff
@@ -2,26 +2,30 @@
     "lodash": "^4.17.20",
     "serve-handler": "^6.1.3",
     "server": "*",
     "stubby": "^5.0.0",
```
As per conversation, see if you can get rid of this completely

psmithshine committed:
remove all-things "Stubby"

psmithshine committed:
Upgrade various Apollo and React packages with code to go with it, add Vitest packages, config, and testing script, improve integration's port resolution, update Readme to reflect integration updates

psmithshine committed:
upgrade to latest MockServiceWorker (msw) and update code to use new API

psmithshine committed:
clarify instructions again for setting up integration env

psmithshine committed:
server codegen upgrades

psmithshine committed:
make yarn install fail if lockfile requires changes

psmithshine committed:
use Zod to validate env.json in server and resolve #15
